FWCore/Framework/interface/Callback.h

```diff
@@ -49,31 +60,249 @@ namespace edm {
+              typename TRecord,       //the record passed in as an argument
+              typename TDecorator     //allows customization using pre/post calls
               = CallbackSimpleDecorator<TRecord>>
     class Callback {
     public:
-      Callback(T* iProd, TFunc iFunc, unsigned int iID, const TDecorator& iDec = TDecorator())
-          : Callback(iProd, std::make_shared<TFunc>(std::move(iFunc)), iID, iDec) {}
+      Callback(T* iProd, TProduceFunc iProduceFunc, unsigned int iID, const TDecorator& iDec = TDecorator())
+          : Callback(iProd,
+                     std::shared_ptr<TAcquireFunc>(),
+                     std::make_shared<TProduceFunc>(std::move(iProduceFunc)),
+                     iID,
+                     iDec) {}
+
+      Callback(T* iProd,
+               TAcquireFunc iAcquireFunc,
+               TProduceFunc iProduceFunc,
+               unsigned int iID,
+               const TDecorator& iDec = TDecorator())
+          : Callback(iProd,
+                     std::make_shared<TAcquireFunc>(std::move(iAcquireFunc)),
+                     std::make_shared<TProduceFunc>(std::move(iProduceFunc)),
+                     iID,
+                     iDec) {}
 
-      Callback(T* iProd, std::shared_ptr<TFunc> iFunc, unsigned int iID, const TDecorator& iDec = TDecorator())
+      Callback(T* iProd,
+               std::shared_ptr<TAcquireFunc> iAcquireFunc,
+               std::shared_ptr<TProduceFunc> iProduceFunc,
+               unsigned int iID,
+               const TDecorator& iDec = TDecorator())
           : proxyData_{},
             producer_(iProd),
             callingContext_(&iProd->description()),
-            function_(std::move(iFunc)),
+            acquireFunction_(std::move(iAcquireFunc)),
+            produceFunction_(std::move(iProduceFunc)),
             id_(iID),
             wasCalledForThisRecord_(false),
             decorator_(iDec) {}
 
-      Callback* clone() { return new Callback(producer_.get(), function_, id_, decorator_); }
+      Callback* clone() { return new Callback(producer_.get(), acquireFunction_, produceFunction_, id_, decorator_); }
 
       Callback(const Callback&) = delete;
       const Callback& operator=(const Callback&) = delete;
 
+      class ESProduceTask : public WaitingTask {
+      public:
+        ESProduceTask(Callback* iCallback,
+                      oneapi::tbb::task_group* iGroup,
+                      ServiceToken const& iServiceToken,
+                      EventSetupRecordImpl const* iRecord,
+                      EventSetupImpl const* iEventSetupImpl)
+            : callback_(iCallback),
+              group_(iGroup),
+              serviceToken_(iServiceToken),
+              record_(iRecord),
+              eventSetupImpl_(iEventSetupImpl) {}
+
+        void execute() final {
+          auto excptr = exceptionPtr();
+          if (!callback_->acquireFunction_) {
+            CMS_SA_ALLOW try {
```

Comment:
You're right. It is not needed here. I will remove CMS_SA_ALLOW. It's harmless, but unnecessary because I used convertException::wrap. Thanks for noticing.